[ROCm] StreamExecutor logic for ROCm platform

whchung opened the pull request:
Please focus on the content in commit https://github.com/tensorflow/tensorflow/pull/20709/commits/33505b8cfa1298653e4954974abcbbadbcb746ca

The pull request contains StreamExecutor logic for ROCm platform. It includes
integration logic with major components on ROCm platform:

- HIP runtime APIs
- rocRAND for RNG

Also included are relevant changes to:

- bazel script to build ROCm StreamExecutor on ROCm platform

Following components in ROCm StreamExecutor are not included and will be filed
as separate pull requests:

- #20712 : rocFFT for FFT algorithms
- #20713 : MIOpen for DNN algorithms
- #20715 : rocBLAS for BLAS algorithms

There are additional changes included in this pull request which are covered
in other pull requests for ROCm. In order to ensure each pull request is
self-contained and builds / runs on both CUDA & ROCm platform they are still
included here.

- #20277 : bazel changes and continuous integration logic
- #20675 : StreamExecutor interface change

Authors:

- Jack Chung: jack.chung@amd.com
- Deven Desai: deven.desai@amd.com
- Johannes M Dieterich: Johannes.Dieterich@amd.com
- Peng Sun: Peng.Sun@amd.com
- Jeffrey Poznanovic: Jeffrey.Poznanovic@amd.com

whchung committed:
[ROCm] StreamExecutor logic for ROCm platform

The commit contains StreamExecutor logic for ROCm platform. It includes
integration logic with major components on ROCm platform:

- HIP runtime APIs
- rocRAND for RNG

Also included are relevant changes to:

- bazel script to build ROCm StreamExecutor on ROCm platform

Authors:

- Jack Chung: jack.chung@amd.com
- Deven Desai: deven.desai@amd.com
- Johannes M Dieterich: Johannes.Dieterich@amd.com
- Peng Sun: Peng.Sun@amd.com
- Jeffrey Poznanovic: Jeffrey.Poznanovic@amd.com

whchung committed:
[ROCm] Interface changes for StreamExecutor to support both CUDA and ROCm

1) StreamInterface::CudaStreamMemberHack()

Despite the fact that StreamExecutor and GPU common runtime are largely
orthogonal, this particular routine in StreamExecutor is used in GPU common
runtime and a couple of other operators. In this commit it's renamed as
StreamInterface::GpuStreamMemberHack() and their call sites are also changed.

2) StreamExecutorInterface::CudaContextHack()

This member is renamed to StramExecutorInterface::GpuContextHack().

Changes introduced in this commit includes:

- some StreamExecutor interfaces and CUDA implementation
- GPU common runtime related to interface changes in StreamExecutor
- operators affected by interface changes in StreamExecutor

jlebar commented on tensorflow/stream_executor/stream.h:
```diff
@@ -628,43 +628,49 @@ class Stream {
                           const dnn::BatchDescriptor &output_dimensions,
-                          DeviceMemory<double> *output_data);
+                          DeviceMemory<double> *output_data,
+                          ScratchAllocator *workspace_allocator = nullptr);
```
Shouldn't these changes not be present in this commit?

whchung replied:
This is made intentionally to make sure subsequent PRs such as #20713 be built properly.

jlebar replied:
Part of the purpose of splitting out the patches is to make subsequent patches easier to read.  If we combine all of the dependent patches in here, then this one is hard to read, defeating most of the purpose.

Can you include the other, split-out commits, in this PR?  Then the top commit in this PR won't contain the commits you split out, and we can just review the top commit.

Alternatively I guess you could just send one mega PR with multiple commits.  I've never done a big review in Github, so I'm not sure which would be easier; I guess it's up to you.  The important thing is that each atomic change is split out and individually documented, because that makes it much easier to review.

whchung replied:
Thanks for the feedback. Let me see how to refactor this PR.

jlebar reviewed (commented):
From the commit message and the code, it looks like some of the changes that we pulled out into separate commits are still present in this change.  Is that intentional?

whchung committed:
[ROCm] bazel build system and continuous integration logic

The commit contains following components to support TensorFlow on ROCm platform

- bazel build system
- continuous integration logic

Authors:

- Jack Chung: jack.chung@amd.com
- Jeffrey Poznanovic: Jeffrey.Poznanovic@amd.com
- Peng Sun: Peng.Sun@amd.com

whchung commented:
@jlebar Based on your comments, I've revised the PR again and only cherry-pick commits in previous PRs which are truly necessary for this particular PR to build.

All relevant changes in this particular PR has been placed in 1 commit:
https://github.com/tensorflow/tensorflow/pull/20709/commits/33505b8cfa1298653e4954974abcbbadbcb746ca

jlebar commented:
cc @chsigg

whchung commented:
@qlzh727 I addressed styling found in "Ubuntu Sanity" check. Could you help restart the test? Thanks.

tensorflowbutler commented:
Nagging Reviewer @chsigg: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
It has been 20 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

qlzh727 commented:
@whchung, please rebase and resolve the merge conflict here if this PR is still alive.

deven-amd commented:
@qlzh727 

This PR is still active. Waiting for PR #20277 to get merged, since that is a per-requisite to getting this PR merged.  Once that PR is merged, we expect to rebase and resolve the merge conflicts.

Thanks

tensorflowbutler commented:
It has been 28 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
